### PR TITLE
Issue 6618: LTS - Perf Improvement - Remove unnecessary pessimistic check during read of SystemJournals.

### LINKED ISSUE #6618: LTS - Perf Improvement - Remove unnecessary pessimistic check during read of SystemJournals
**Describe the bug**
Currently there are redundant `exists `and `getInfo` calls which are unnecessary during read of SystemJournals. They should be removed and replaced with optimistic check.

**To Reproduce**
During perf tests we found high number of Head requests related to SystemJournals

Code inspection shows pessimistic checks in following places

- https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java#L570-L576
- https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java#L570-L576

**Screenshots**
None

**Additional information**
None

### PR DESCRIPTION
**Pull request by sachin-j-joshi**
Signed-off-by: Sachin Joshi <sachin.joshi@emc.com>

**Change log description**  
Remove unnecessary pessimistic check during read of SystemJournals. Instead use optimistic logic and handle failures. 

**Purpose of the change**  
Fixes #6618

**What the code does**  
Remove unnecessary pessimistic check during read of SystemJournals. Instead use optimistic logic and handle failures. 

**How to verify it**  
All tests should pass.

### CODE REVIEW

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java [hunk 1]
```diff
@@ -568,31 +568,22 @@ public void initialize() throws Exception {
...
-                                // Step 3: Read contents.
-                                .thenComposeAsync(v -> getContents(snapshotFileName), executor)
-                                // Step 4: Deserialize and return.
-                                .thenApplyAsync(contents -> readSnapshotRecord(snapshotInfo, contents), executor);
+                        // Step 2: Validate. and Read contents.
+                        return getContents(snapshotFileName, false)
+                                // Step 3: Deserialize and return.
+                                .thenApplyAsync(contents -> readSnapshotRecord(snapshotInfo, contents), executor)
+                                .exceptionally( e -> {
+                                    throw new CompletionException(new IllegalStateException(
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:576]: Whatever exception may come from `readSnapshotRecord` will be caught by this block and will be converted into this `IllegalStateException`, even though the problem may be unrelated to the check of chunk existence. Is this conceptually correct? Can this cause any problem downstream?

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java [hunk 2]
```diff
@@ -568,31 +568,22 @@ public void initialize() throws Exception {
...
+                                            String.format("Chunk pointed by SnapshotInfo must exist. Missing chunk %s", snapshotFileName),
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:577]: Please, can you add a test by inducing an exception being thrown from `readSnapshotRecord` to see if it is being handled properly?

**Review by RaulGracia — CHANGES_REQUESTED** [review]
I left some minor comments. Also, what is the expectation of improvement with this change? Avoid unnecessary GETs to the underlying storage while scanning journals, right?

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java (hunk 2 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:577]: We are testing all sorts of failures pretty extensively in 
https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/SystemJournalOperationsTests.java

These tests already cover the scenario.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java (hunk 1 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:576]: IllegalStateException is the correct behavior. In ability to read already created and validated snapshot is a critical error that prevents SLTS from boot.
readSnapshotRecord we retry several times before throwing exception.

Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:576]: However, the message certainly needs to be more generic

Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:576]: fixed.

**Comment by sachin-j-joshi** [comment]
@RaulGracia Addressed the comments. Thanks

**Comment by RaulGracia** [comment]
I see system tests failing with this branch: `test-pravega-continuous-1/2082`
Please, download the logs and let us know the cause for the test failure and whether it is related or not with these changes.

**Review by RaulGracia — CHANGES_REQUESTED** [review]
Blocking for safety until system tests are consistently green.

**Comment by RaulGracia** [comment]
@sachin-j-joshi I assume that the failure I pointed out of system tests is a false positive unrelated to this change, right?

**Pull request merged**